### PR TITLE
Solution

### PR DESCRIPTION
**Pull request by nikhilm**
useful to have node or something node-like to test by just running `node answer.js`

### CODE REVIEW

File: week001/answer.js [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+  // arguments.
+  retfn.valueOf = function() {
+    return retfn;
+  };
+  retfn.toString = function() { return sum; }
+  return retfn;
+}
+
+exports.add = add;
+assert.ok(add(2, 5) == 7);
```
Comment (nikhilm) [inline week001/answer.js]: note that using the form assert.equals(add(2,5), 7) will fail

File: week001/answer.js [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+assert.ok(add(2)(5) == 7);
```
Comment (nikhilm) [inline week001/answer.js:29]: note that using the form assert.equals(add(2)(5), 7) will fail

**Comment by cdax** [comment]
Dammit, your redefinition of `valueOf()` was clever. Makes my own solution look unnecessarily verbose even though it isn't.

**Comment by cdax** [comment]
Let me know if you've got challenge suggestions for next week or the week after.

**Comment by nikhilm** [comment]
Thanks! I didn't know about valueOf before either. For this problem it was pretty obvious that I had to use some form of arguments array abuse and return the add function from within itself, but I didn't know how to get it to return different things based on context. So I looked into if like C++ allows you to overload the callable operator(), JS allows functions to do something similar. It doesn't, but I stumbled upon arithmetic operator overloading (which is quite grotesque) - http://stackoverflow.com/questions/1634341/overloading-arithmetic-operators-in-javascript that gave me the valueOf idea.

This challenge was based entirely on abusing JS and you'd never use this sort of code in production :P so it would be nice if the next challenge wasn't like this.

**Comment by cdax** [comment]
Even so, I'm looking at it this way: I'll use the solution post to explain currying. The [original problem](https://github.com/h5bp/Front-end-Developer-Interview-Questions#coding-questions) from the interview problems repo requires you to pass just two tests:

`add(2, 5);` and
`add(2)(5);`

...which is almost certainly a check to see if the candidate's aware of currying. Also, for this simple(r) case, no `toString`/`valueOf` trickery would be required. 

Anyway, doing this hopefully makes this first challenge practically useful for readers.

**Comment by nikhilm** [comment]
Ok that makes sense.

**Pull request merged**